File: scripts/launchAnalysis.pl
#!/usr/bin/perl

###################################
## Code to launch the analysis   ##
###################################

#--------------------------------------------------------------
# Francesco Santanastasio  <francesco.santanastasio@cern.ch>
#--------------------------------------------------------------

print "Starting...\n";

use Time::Local;
use Getopt::Std;

## input info

my $inputList;
my $outputDir;
my $treename;
my $cutfile;

getopts('h:i:o:n:c:');

if(!$opt_i) {help();}
if(!$opt_o) {help();}
if(!$opt_n) {help();}
if(!$opt_c) {help();}

if($opt_h) {help();}
if($opt_i) {$inputList = $opt_i;}
if($opt_o) {$outputDir = $opt_o;}
if($opt_n) {$treename = $opt_n;}
if($opt_c) {$cutfile = $opt_c;}

system "mkdir -p $outputDir";
system "cp $cutfile $outputDir/";

open CUTFILE, $cutfile or die $!;

my $found_json = 0;
my $jsonFile = "";

while (<CUTFILE>) {
    my $line = trim ($_);
    my $first_four_characters = substr($line,0,4);
    if (lc($first_four_characters) eq "json" )
    {	
	
	@split_line = split (" ", $line) ;
	$number_of_elements = $#split_line + 1;

	if ( $found_json ) 
	{
	    print "ERROR: Please specify only one JSON file in your cut file!\n";
	    exit 0;
 	}

	if ( $#split_line != 1 ) 
	{
	    print "ERROR: Too many ($number_of_elements) elements in the line.  Line reads:\n";
	    print "  $line\n";
	    print "  Format in cut file should be:\n";
	    print "  JSON <json file full path>\n";
	    exit 0;
	}

	$jsonFile = @split_line[1];

	open (JSONFILE, "<$jsonFile") || die ("ERROR: JSON file $jsonFile specified in $cutfile does not exist");
	close (JSONFILE );
   
	$found_json=1
    }    
} 

close (CUTFILE);

if ( $found_json ) {
    print  "cp $jsonFile $outputDir/\n";
    system "cp $jsonFile $outputDir/";
}

open (INPUTLIST, "<$inputList") || die ("...error reading file $inputList $!");
@inputList = <INPUTLIST>;
#print @inputList;
close(INPUTLIST);



open (FILE, "ls -l src/analysisClass.C |") || die ("...error reading the file src/analysisClass.C $!");
$analysisClassFull = <FILE>;
close(FILE);

## 1st split
my @array = split(/\s+/ , $analysisClassFull );
#print "$analysisClassFull\n";
#print "$array[10]\n";

if ( scalar(@array) != 11 ) 
{die ("...error src/analysisClass.C is not a symbolic link");}

## 2nd split
my @codenameC = split(/\// , $array[10] );
#print "@codenameC\n";
#print "$codenameC[scalar(@codenameC)-1]\n";

## 3rd split
my ($codename,$EXT) = split(/\./ , $codenameC[scalar(@codenameC)-1] );
#print "$codename\n";

for $line(@inputList)
{
    chomp($line);

    my @array1 = split(/\// , $line );
    my ($dataset,$EXT) = split(/\./ , $array1[scalar(@array1)-1] );

    print "./main $line $cutfile $treename $outputDir/$codename\_\_\_$dataset $outputDir/$codename\_\_\_$dataset \n";
    system "./main $line $cutfile $treename $outputDir/$codename\_\_\_$dataset $outputDir/$codename\_\_\_$dataset";
}

#---------------------------------------------------------#

sub help(){
    print "Usage: ./script/launchAnalysis.pl -i <inputList> -c <cutfile> -n <treename> -o <outputDir> [-h <help?>] \n";
    print "Example: ./launchAnalysis.pl -i /home/santanas/Workspace/Leptoquarks/rootNtupleAnalyzer/config/inputListAllCurrent.txt -c cutFileExample.txt -n RootTupleMaker -o /home/santanas/Workspace/Leptoquarks/rootNtupleAnalyzer/data/output\n";
    print "Options:\n";
    print "-i <inputList>:      choose the file containing all the input lists for the analysis\n";
    print "-n <treename>:       choose the name of the TTree of the .root files you want to analyze\n";
    print "-c <cutfile>:        choose the name of the file with the analysis cuts\n";
    print "-o <outputDir>:      choose the output directory where the .root files will be stored\n";
    print "-h <yes> :           to print the help \n";
    die "please, try again...\n";
}

# Perl trim function to remove whitespace from the start and end of the string
sub trim($)
{
	my $string = shift;
	$string =~ s/^\s+//;
	$string =~ s/\s+$//;
	return $string;
}

File: scripts/launchAnalysis.py
#!/usr/bin/python
# If you're processing a smaller amount of data (as in the 2F region of the closure test) it's faster to use this, so I updated it. -Emma
#----------------------------------------------------------------
# Imports
#----------------------------------------------------------------

import optparse
import sys
import os
import multiprocessing
from multiprocessing import Pool

#----------------------------------------------------------------
# Define functions
#----------------------------------------------------------------

def printAndExecute ( command ): 
    print (command)
    os.system ( command ) 

#----------------------------------------------------------------
# Get user options
#----------------------------------------------------------------

parser = optparse.OptionParser(description='Launch the analysis')
parser.add_option ("--input_list", "-i" , dest = "input_list", type=str, help="Text file containing paths of input lists with .root files for analysis")
parser.add_option ("--output_dir", "-o" , dest = "output_dir", type=str, help="Directory where output will be stored" ) 
parser.add_option ("--tree_name" , "-n" , dest = "tree_name" , type=str, help="Name of TTree in .root files to be analyzed")
parser.add_option ("--cut_file"  , "-c" , dest = "cut_file"  , type=str, help="Text file containing cut values" ) 
parser.add_option ("--ncores"    , "-p" , dest = "ncores"    , type=int, help="Number of processor cores to be used to run the job" )
parser.add_option ("--exe"       , "-e" , dest = "execut"    , type=str, help="name of exec" )
parser.add_option ("--code_name" , "-k" , dest = "code_name" , type=str, help="code name" )

(options, args) = parser.parse_args()

if not options.input_list : parser.print_help(); sys.exit()
if not options.output_dir : parser.print_help(); sys.exit()
if not options.tree_name  : parser.print_help(); sys.exit()
if not options.cut_file   : parser.print_help(); sys.exit()
if not options.code_name  : parser.print_help(); sys.exit()

ncores = multiprocessing.cpu_count()
if options.ncores :
    ncores = options.ncores

executable = "main"
if options.execut:
  executable = options.execut

#----------------------------------------------------------------
# Bail if the cut file doesn't exist:
#----------------------------------------------------------------

if not os.path.isfile ( options.cut_file ) :
    print ("ERROR.  Cut file does not exist: " + options.cut_file)
    sys.exit() 

#----------------------------------------------------------------
# Bail if the input list doesn't exist:
#----------------------------------------------------------------

if not os.path.isfile ( options.input_list ) : 
    print ("ERROR.  Input list does not exist: " + options.input_list)
    sys.exit() 

#----------------------------------------------------------------
# Make the output directory if it doesn't exist
#----------------------------------------------------------------

if not os.path.isdir ( options.output_dir ) : 
    print ("Making directory: " + options.output_dir)
    #os.mkdir ( options.output_dir )
    os.makedirs ( options.output_dir ) 

#----------------------------------------------------------------
# Can you find a JSON file? If so, save it
#----------------------------------------------------------------

cut_file = open ( options.cut_file, "r" ) 
found_json = False
for line in cut_file:
    line = line.split()
    if line[0:4] == "JSON":

        json_file_path = ""

        if found_json:
            print ("ERROR.  Cut file contains multiple JSON files")
            print ("   Cut file is: " + options.cut_file)
            sys.exit()
        elif len (line.split() ) != 2 : 
            print ("ERROR.  Cut file contains bad syntax about JSON file.")
            print ("   Cut file is: " + options.cut_file)
            print ("   Line should read: JSON <path to JSON file>")
            print ("   Line reads      : " + line)
            sys.exit()
        elif not os.path.isfile ( line.split()[1] ) :
            print ("ERROR. JSON file in cut file does not exist")
            print ("   Cut file is:  " + options.cut_file)
            print ("   JSON path is: " + line.split()[1])
            sys.exit() 
        else : 
            command = "cp " + line.split()[1] + " " + options.output_dir 
            print (command)
            os.system ( command ) 
            found_json = True

cut_file.close() 

#----------------------------------------------------------------
# Is the analysisClass.C file a symbolic link?
# If it is, what is the name of the real .C file?
#----------------------------------------------------------------
#
#code_name = ""
#if not os.path.islink ("src/analysisClass.C" ) :
#    print "ERROR.  src/analysisClass.C is not a symbolic link!" 
#    sys.exit() 
#else : 
#    code_name = os.path.realpath ( "src/analysisClass.C" ).split("/")[-1][:-2]

#----------------------------------------------------------------
# Loop over the input list and get a command for each file
#----------------------------------------------------------------

input_list = open ( options.input_list, "r" ) 
command_list = []

for line in input_list:
    if line.startswith('#'):
      continue
    dataset = line.strip().split("/")[-1][:-4]
    output_file_name = options.output_dir + "/" + options.code_name + "___" + dataset    
    command = "./" + executable
    command += " " + line.strip() + " " + options.cut_file + " " + options.tree_name + " " + output_file_name + " " + output_file_name
    command_list.append ( command ) 

#----------------------------------------------------------------
# Execute using all available CPUs.  Quit on Ctrl-C.
#----------------------------------------------------------------

pool = Pool ( ncores ) 
try:
    pool.map_async( printAndExecute, command_list ).get(99999999)
except KeyboardInterrupt:
    print ("\n\nCtrl-C detected: Bailing.") 
    pool.terminate()
    sys.exit(1) 
